### PR TITLE
Add support for JOIN USING statements

### PR DESCRIPTION
**Pull request by marregui**
Addressing https://github.com/crate/crate/issues/9551

### CODE REVIEW

**Comment by marregui** [comment]
Maybe I can get some early feedback to see if I completely blew this, before I add specific tests.

**Comment by mfussenegger** [comment]
> 
> 
> Maybe I can get some early feedback to see if I completely blew this, before I add specific tests.

Looks good so far

File: sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java [hunk 1]
```diff
@@ -1190,7 +1189,9 @@ public Node visitJoinRelation(SqlBaseParser.JoinRelationContext context) {
                 criteria = new JoinOn((Expression) visit(context.joinCriteria().booleanExpression()));
             } else if (context.joinCriteria().USING() != null) {
                 List<String> columns = identsToStrings(context.joinCriteria().ident());
-                criteria = new JoinUsing(columns);
+                Table leftTable = (Table) left;
```
Comment (kovrus) [inline sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java]: Is it really safe to cast to `Table` here? What if `left` or `right` will be a sub query?

Comment (marregui) [reply sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java]: Not safe, I will now add tests, like the one https://github.com/crate/crate/issues/9551 , which will highlight in catastrophic form exactly this :)

File: docs/sql/statements/select.rst [hunk 2]
```diff
@@ -159,7 +159,7 @@ Joined relation
 A ``joined_relation`` is a relation which joins two relations together. See
 :ref:`sql_dql_joins` ::
 
-    relation { , | join_type JOIN } relation [ ON join_condition ]
+    relation { , | join_type JOIN } relation {[ ON join_condition ] | [USING (col_names)]}
```
Comment (kovrus) [inline docs/sql/statements/select.rst]: `ON` and `USING` should be optional, so not sure if it is correct to put them into curly braces. Maybe something like this would work?
```suggestion
    relation { , | join_type JOIN } relation [ ON join_condition | USING (column_ident [, ...]) ]
```

File: docs/appendices/release-notes/unreleased.rst [hunk 3]
```diff
@@ -65,6 +65,10 @@ None
 Changes
 =======
 
+- Added support for ``JOIN USING``, e.g. select * from t1 join t2 using(col1, col2),
```
Comment (mfussenegger) [inline docs/appendices/release-notes/unreleased.rst]:
```suggestion
- Added support for ``JOIN USING``, e.g. ``SELECT * FROM t1 JOIN t2 USING (col)``,
```

File: docs/appendices/release-notes/unreleased.rst [hunk 4]
```diff
@@ -65,6 +65,10 @@ None
...
+  a more convenient alternative to ``JOIN ON``, when we know both tables have the
```
Comment (mfussenegger) [inline docs/appendices/release-notes/unreleased.rst]:
```suggestion
  an alternative to ``JOIN ON``, when the column name(s) are the same in both relations.
```

File: docs/sql/statements/select.rst (hunk 2 above)
Comment (mfussenegger) [inline docs/sql/statements/select.rst]: I think it is:
```suggestion
    relation { , | join_type JOIN } relation [ { ON join_condition  |  USING (col_names) } ]
```

`[ .. ]`for the optional indication, `{ a | b | c }` to denote different options.

File: docs/sql/statements/select.rst [hunk 5]
```diff
@@ -171,6 +171,11 @@ A ``joined_relation`` is a relation which joins two relations together. See
   The join_condition is not applicable for joins of type CROSS and must
   have a returning value of type ``boolean``.
 
+:col_names:
+  A comma separated list of column names. The joining tables are both expected
```
Comment (mfussenegger) [inline docs/sql/statements/select.rst]:
```suggestion
  A comma separated list of column names. The relations joined need to contain the columns specified.
```

File: sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java [hunk 6]
```diff
@@ -480,8 +480,8 @@ public Node visitRestore(SqlBaseParser.RestoreContext context) {
                 extractGenericProperties(context.withProperties()));
         }
         return new RestoreSnapshot(getQualifiedName(context.qname()),
-            visitCollection(context.tableWithPartitions().tableWithPartition(), Table.class),
-            extractGenericProperties(context.withProperties()));
+                                   visitCollection(context.tableWithPartitions().tableWithPartition(), Table.class),
```
Comment (mfussenegger) [inline sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java]: Please revert unrelated indentation changes

File: sql/src/test/java/io/crate/analyze/SelectStatementAnalyzerTest.java [hunk 7]
```diff
@@ -161,15 +161,15 @@ public void testOrderedSelect() throws Exception {
...
         AnalyzedRelation relation =  analyze("select * from sys.nodes where port['http'] = -400");
         Function whereClause = (Function) relation.where().query();
         Symbol symbol = whereClause.arguments().get(1);
-        assertThat(((Literal) symbol).value(), is(-400));
+        assertThat(((Literal<?>) symbol).value(), is(-400));
     }
 
     @Test
-    public void testSimpleSelect() throws Exception {
+    public void testSimpleSelect() {
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/analyze/SelectStatementAnalyzerTest.java]: Please also revert these removals. It causes noise that is unrelated to the change of the PR.

 I know the IDE bugs you about it, but for tests it doesn't matter and it is usually annoying to re-add it once any of the methods tell the IDE that they throw something.

Comment (marregui) [reply sql/src/test/java/io/crate/analyze/SelectStatementAnalyzerTest.java]: My bad, I was getting blinded by the yellow column in the IDE, which hid some red.
I have seen people treat warnings as errors in other lives.

**Comment by marregui** [comment]
retest this please

File: sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java [hunk 8]
```diff
@@ -1190,14 +1190,35 @@ public Node visitJoinRelation(SqlBaseParser.JoinRelationContext context) {
...
-                criteria = new JoinUsing(columns);
+                criteria = new JoinUsing(extractQualifiedName(left), extractQualifiedName(right), columns);
             } else {
                 throw new IllegalArgumentException("Unsupported join criteria");
             }
         }
         return new Join(getJoinType(context.joinType()), left, right, Optional.of(criteria));
     }
 
+    private static QualifiedName extractQualifiedName(Relation rel) {
```
Comment (mfussenegger) [inline sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java]: This instanceof handling in this method might be easy to forget to update in case we add new relations.

I like the solution you had in the first commit better. Was there a reason why you choose to go with this instead?

Comment (marregui) [reply sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java]: I like the original commit better too.

File: sql/src/main/java/io/crate/analyze/relations/RelationAnalyzer.java [hunk 9]
```diff
@@ -253,7 +255,7 @@ protected AnalyzedRelation visitJoin(Join node, StatementAnalysisContext stateme
         Symbol joinCondition = null;
         if (optCriteria.isPresent()) {
             JoinCriteria joinCriteria = optCriteria.get();
-            if (joinCriteria instanceof JoinOn) {
+            if (joinCriteria instanceof JoinOn || joinCriteria instanceof JoinUsing) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/relations/RelationAnalyzer.java]: How about something like the following to avoid the duplicate instanceof:

```
diff --git a/sql/src/main/java/io/crate/analyze/relations/RelationAnalyzer.java b/sql/src/main/java/io/crate/analyze/relations/RelationAnalyzer.java
index 80f718714f..8e3c5da7cf 100644
--- a/sql/src/main/java/io/crate/analyze/relations/RelationAnalyzer.java
+++ b/sql/src/main/java/io/crate/analyze/relations/RelationAnalyzer.java
@@ -107,7 +107,6 @@ import javax.annotation.Nullable;
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
-import java.util.Iterator;
 import java.util.List;
 import java.util.Locale;
 import java.util.Map;
@@ -247,48 +246,45 @@ public class RelationAnalyzer extends DefaultTraversalVisitor<AnalyzedRelation,
 
     @Override
     protected AnalyzedRelation visitJoin(Join node, StatementAnalysisContext statementContext) {
-        node.getLeft().accept(this, statementContext);
-        node.getRight().accept(this, statementContext);
+        var leftRel = node.getLeft().accept(this, statementContext);
+        var rightRel = node.getRight().accept(this, statementContext);
 
         RelationAnalysisContext relationContext = statementContext.currentRelationContext();
         Optional<JoinCriteria> optCriteria = node.getCriteria();
         Symbol joinCondition = null;
         if (optCriteria.isPresent()) {
             JoinCriteria joinCriteria = optCriteria.get();
-            if (joinCriteria instanceof JoinOn || joinCriteria instanceof JoinUsing) {
-                final CoordinatorTxnCtx coordinatorTxnCtx = statementContext.transactionContext();
-                ExpressionAnalyzer expressionAnalyzer = new ExpressionAnalyzer(
-                    functions,
-                    coordinatorTxnCtx,
-                    statementContext.convertParamFunction(),
-                    new FullQualifiedNameFieldProvider(
-                        relationContext.sources(),
-                        relationContext.parentSources(),
-                        coordinatorTxnCtx.sessionContext().searchPath().currentSchema()),
-                    new SubqueryAnalyzer(this, statementContext));
-                Expression expr;
-                if (joinCriteria instanceof JoinOn) {
-                    expr = ((JoinOn) joinCriteria).getExpression();
-                } else {
-                    Iterator<QualifiedName> rels = relationContext.sources().keySet().iterator();
-                    QualifiedName left = rels.hasNext() ? rels.next() : null;
-                    QualifiedName right = rels.hasNext() ? rels.next() : null;
-                    if (null == left || null == right) {
-                        throw new IllegalStateException("joining tables are undefined");
-                    }
-                    expr = JoinUsing.toExpression(left, right, ((JoinUsing) joinCriteria).getColumns());
-                }
-                try {
-                    joinCondition = expressionAnalyzer.convert(expr, relationContext.expressionAnalysisContext());
-                } catch (RelationUnknown e) {
-                    throw new RelationValidationException(e.getTableIdents(),
-                        String.format(Locale.ENGLISH,
-                        "missing FROM-clause entry for relation '%s'", e.getTableIdents()));
-                }
+            final CoordinatorTxnCtx coordinatorTxnCtx = statementContext.transactionContext();
+            ExpressionAnalyzer expressionAnalyzer = new ExpressionAnalyzer(
+                functions,
+                coordinatorTxnCtx,
+                statementContext.convertParamFunction(),
+                new FullQualifiedNameFieldProvider(
+                    relationContext.sources(),
+                    relationContext.parentSources(),
+                    coordinatorTxnCtx.sessionContext().searchPath().currentSchema()),
+                new SubqueryAnalyzer(this, statementContext)
+            );
+            final Expression expr;
+            if (joinCriteria instanceof JoinOn) {
+                expr = ((JoinOn) joinCriteria).getExpression();
+            } else if (joinCriteria instanceof JoinUsing) {
+                expr = JoinUsing.toExpression(
+                    leftRel.getQualifiedName(),
+                    rightRel.getQualifiedName(),
+                    ((JoinUsing) joinCriteria).getColumns()
+                );
             } else {
                 throw new UnsupportedOperationException(String.format(Locale.ENGLISH, "join criteria %s not supported",
                     joinCriteria.getClass().getSimpleName()));
             }
+            try {
+                joinCondition = expressionAnalyzer.convert(expr, relationContext.expressionAnalysisContext());
+            } catch (RelationUnknown e) {
+                throw new RelationValidationException(e.getTableIdents(),
+                    String.format(Locale.ENGLISH,
+                    "missing FROM-clause entry for relation '%s'", e.getTableIdents()));
+            }
         }
 
         relationContext.addJoinType(JoinType.values()[node.getType().ordinal()], joinCondition);

```

and to avoid the left / right null cases.
